### PR TITLE
enh(Licenses) : Implement a method that calculates the number of days remaining until the license expiration date

### PR DESCRIPTION
**Pull request by Med1233**
## Description

Implement a method that calculates the number of days remaining until the license expiration date

## Type of change

- [x] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [x] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 21.10.x
- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [x] 23.10.x (master)

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: centreon/packages/ui/src/utils/calculateDaysUntilLicenseExpiration.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  response,
+  module
+}: Props): number => {
+  const getExpirationDate = pipe(
+    path(['result', 'module', 'entities']),
+    find(propEq('id', module)),
+    path(['license', 'expiration_date'])
+  ) as (data) => string;
+
+  const currentDate = new Date();
```
Comment (Thebarda) [inline centreon/packages/ui/src/utils/calculateDaysUntilLicenseExpiration.ts:17]: Use dayjs for date management

Comment (bdauria) [reply centreon/packages/ui/src/utils/calculateDaysUntilLicenseExpiration.ts:17]: Should we add a unit test for this util or do we handle the coverage at the component level?

Comment (Med1233) [reply centreon/packages/ui/src/utils/calculateDaysUntilLicenseExpiration.ts]: my current implementation involves an initial test of the function on a standalone, generic component (an empty div), devoid of any additional behaviors.  and then proceed to test it again on every page within each module.

File: centreon/packages/ui/src/utils/useLicenseExpirationWarning.cypress.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    Component: <TestWithQueryProvider />
+  });
+};
+
+const currentDate = dayjs();
+
+describe('License', () => {
+  beforeEach(initialize);
+
+  it('does not display a warning message when the license will expire in more than 15 days', () => {
```
Comment (bdauria) [inline centreon/packages/ui/src/utils/useLicenseExpirationWarning.cypress.spec.tsx:121]:
```suggestion
  it('does not display any warning message when the license expires in more than 15 days from the current date', () => {
```

File: centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 3]
```diff
@@ -17978,4 +17978,7 @@ msgstr "Une erreur est survenue lors de la récupération des données de métri
 
 
 msgid "Invalid metric format"
-msgstr "Format de métrique invalide"
\ No newline at end of file
+msgstr "Format de métrique invalide"
+
+msgid "License {{module}} will expire in {{expirationDate}} days"
```
Comment (smau-centreon) [inline centreon/lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:17983]:
```suggestion
msgid "The {{module}} license will expire in {{expirationDate}} days"
```

File: centreon/packages/ui/src/utils/useLicenseExpirationWarning.cypress.spec.tsx (hunk 2 above)
Comment (smau-centreon) [inline centreon/packages/ui/src/utils/useLicenseExpirationWarning.cypress.spec.tsx:121]:
```suggestion
  it('does not display a warning message when the license expires in more than 15 days', () => {
```

File: centreon/packages/ui/src/utils/useLicenseExpirationWarning.cypress.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    mockRequest({ expirationDate: currentDate.add(20, 'day') });
+
+    cy.waitForRequest('@getLicenseInformations');
+
+    cy.findByText(
+      labelLicenseWarning('centreon-autodiscovery-server', 20)
+    ).should('not.exist');
+  });
+  it('displays a warning message when the license will expire until less than 15 days', () => {
```
Comment (smau-centreon) [inline centreon/packages/ui/src/utils/useLicenseExpirationWarning.cypress.spec.tsx:130]:
```suggestion
  it('displays a warning message when the license expires within 15 days', () => {
```

**Review by smau-centreon — CHANGES_REQUESTED** [review]

File: centreon/packages/ui/src/utils/useLicenseExpirationWarning.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        return;
+      }
+
+      const daysUntilExpiration = dayjs(expirationDate).diff(
+        currentDate,
+        'day'
+      );
+
+      if (lt(daysUntilExpiration, 15)) {
+        showWarningMessage(labelLicenseWarning(module, daysUntilExpiration));
```
Comment (Thebarda) [inline centreon/packages/ui/src/utils/useLicenseExpirationWarning.ts:46]: This label is not translated

**Review by bdauria — DISMISSED** [review]

**Review by Thebarda — DISMISSED** [review]

**Review by smau-centreon — APPROVED** [review]

**Review by smau-centreon — APPROVED** [review]

**Review by Thebarda — APPROVED** [review]

**Pull request merged**